### PR TITLE
rust: add `NoWaitLock`

### PR DESCRIPTION
**Pull request by wedsonaf**
This is a lock that never waits, that is, if it is owned by another
thread/CPU, then attempts to acquire it will fail (instead of, for
example, blocking the caller).

We'll use this to store `Waker` instances in futures and to fetch them
when reactors need to use them.

Signed-off-by: Wedson Almeida Filho <wedsonaf@google.com>

### CODE REVIEW

**Comment by alex** [comment]
In other contexts this is often seen as `try_lock` method on a traditional lock, is there a reason you went with the dedicated type? Is it more consistent with kernel idioms?

**Comment by wedsonaf** [comment]
> In other contexts this is often seen as `try_lock` method on a traditional lock, is there a reason you went with the dedicated type? Is it more consistent with kernel idioms?

Kernel locks require pinning, unsafe init, and lock classes (for lockdep, which means macros to define and pass them as arguments). This avoids all the ergonomics issues, is smaller, and given that it doesn't ever wait, it doesn't need to participate in lockdep.

File: rust/kernel/sync/nowait.rs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+pub struct NoWaitLockGuard<'a, T: ?Sized> {
+    lock: &'a NoWaitLock<T>,
+}
+
+impl<T: ?Sized> core::ops::Deref for NoWaitLockGuard<'_, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        // SAFETY: The type invariant guarantees that only the owner has an instance of the guard,
+        // so the oowner is the only one that can call this function.
```
Comment (PeterWrighten) [inline rust/kernel/sync/nowait.rs]: The code comment in line 107 has a spelling mistake: 'owner'

**Review by PeterWrighten — COMMENTED** [review]
I found a spelling mistake of code comment in line 107 . As the rest of all, I consider it is a pretty great job!

File: rust/kernel/sync/nowait.rs (hunk 1 above)
Comment (wedsonaf) [reply rust/kernel/sync/nowait.rs]: Fixed. Thanks for spotting this!

**Comment by liuw** [comment]
> We'll use this to store Waker instances in futures and to fetch them
> when reactors need to use them.

So async is going to be used in this project? I guess you will also provide a runtime?

**Comment by wedsonaf** [comment]
> So async is going to be used in this project? I guess you will also provide a runtime?

This project is going to allow async code to be written in modules, yes.

We will provide a runtime.

**Comment by wedsonaf** [comment]
I updated the code to now indicate whether there was contention while the lock was held.

We need this information to determine if we need to immediately trigger a wake up when updating a waker.

File: rust/kernel/sync/nowait.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+//! A lock that never waits.
+
+use core::cell::UnsafeCell;
+use core::sync::atomic::{AtomicU8, Ordering};
+
+const LOCKED: u8 = 1;
+const CONTENDED: u8 = 2;
+
+/// A lock that only offers a [`NoWaitLock::try_lock`] method.
```
Comment (bjorn3) [inline rust/kernel/sync/nowait.rs]: Maybe only show `try_lock`, but still link to the full method?

**Review by bjorn3 — APPROVED** [review]

File: rust/kernel/sync/nowait.rs (hunk 2 above)
Comment (wedsonaf) [reply rust/kernel/sync/nowait.rs]: Yes, it looks better as you suggest.

**Pull request merged**